activerecord/lib/active_record/internal_metadata.rb

```diff
@@ -0,0 +1,56 @@
+require 'active_record/scoping/default'
+require 'active_record/scoping/named'
+
+module ActiveRecord
+  # This class is used to create a table that keeps track of values and keys such
+  # as which environment migrations were run in.
+  class InternalMetadata < ActiveRecord::Base
+    class << self
+      def primary_key
+        "key"
+      end
+
+      def table_name
+        "#{table_name_prefix}#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
+      end
+
+      def index_name
+        "#{table_name_prefix}unique_#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
+      end
+
+      def store(hash)
+        hash.each do |key, value|
+          first_or_initialize(key: key.to_s).update_attributes!(value: value)
+        end
+      end
```

Comment:
Yeah, I think if this were public API that would be more of a concern. For
internal, I think [] and []= are fine.

That said, I don't feel particularly strongly on it, so I'll leave it up to
you since you did all the legwork on this.

On Thu, Jan 7, 2016 at 5:12 PM Matthew Draper <notifications@github.com>
wrote:

> In activerecord/lib/active_record/internal_metadata.rb
> <https://github.com/rails/rails/pull/22967#discussion_r49144499>:
>
> > +        "key"
> > +      end
> > +
> > +      def table_name
> > +        "#{table_name_prefix}#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
> > +      end
> > +
> > +      def index_name
> > +        "#{table_name_prefix}unique_#{ActiveRecord::Base.internal_metadata_table_name}#{table_name_suffix}"
> > +      end
> > +
> > +      def store(hash)
> > +        hash.each do |key, value|
> > +          first_or_initialize(key: key.to_s).update_attributes!(value: value)
> > +        end
> > +      end
>
> That's fair. To me, absent a more comprehensive suite of hash-like
> methods, []/[]= are just the "right" way of spelling "get/set value using
> a key", though.
>
> And this is so far away from anything a user would ever touch, it seems
> fine to go with whatever makes our code look neatest.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/rails/rails/pull/22967/files#r49144499>.
>